crates/runtime/src/instance.rs

```diff
@@ -51,23 +51,27 @@ pub(crate) struct Instance {
     offsets: VMOffsets,
 
     /// WebAssembly linear memory data.
-    memories: BoxedSlice<DefinedMemoryIndex, Box<dyn RuntimeLinearMemory>>,
+    memories: PrimaryMap<DefinedMemoryIndex, Memory>,
 
     /// WebAssembly table data.
-    tables: BoxedSlice<DefinedTableIndex, Table>,
+    tables: PrimaryMap<DefinedTableIndex, Table>,
 
-    /// Passive elements in this instantiation. As `elem.drop`s happen, these
-    /// entries get removed. A missing entry is considered equivalent to an
-    /// empty slice.
-    passive_elements: RefCell<HashMap<ElemIndex, Box<[*mut VMCallerCheckedAnyfunc]>>>,
+    /// Stores the dropped passive element segments in this instantiation by index.
+    /// If the index is present in the set, the segment has been dropped.
+    dropped_elements: RefCell<EntitySet<ElemIndex>>,
 
-    /// Passive data segments from our module. As `data.drop`s happen, entries
-    /// get removed. A missing entry is considered equivalent to an empty slice.
-    passive_data: RefCell<HashMap<DataIndex, Arc<[u8]>>>,
+    /// Stores the dropped passive data segments in this instantiation by index.
+    /// If the index is present in the set, the segment has been dropped.
+    dropped_data: RefCell<EntitySet<DataIndex>>,
 
     /// Hosts can store arbitrary per-instance information here.
     host_state: Box<dyn Any>,
 
+    /// Stores guard page faults in memory relating to the instance.
+    /// This is used for the pooling allocator with uffd enabled on Linux.
+    #[cfg(all(feature = "uffd", target_os = "linux"))]
+    guard_page_faults: RefCell<Vec<(*mut u8, usize, unsafe fn(*mut u8, usize) -> bool)>>,
```

Comment:
I can expand the comment to point at higher-level overview documentation on how uffd works.

Specifically this is a workaround to an issue we were seeing in Lucet on more recent Linux kernels: previously waking a faulting thread without zeroing or copying the page data resulted in a `SIGBUS`, which Lucet relied on to signal an out-of-bounds memory access and thus trapped accordingly.

Newer Linux kernels (5.7+) now retry the fault ad infinitum, resulting in a busy hang in Lucet when uffd is used.

To fix this when implementing uffd in Wasmtime, Wasmtime records faults to current "guard" pages (i.e. pages that *should not* be accessible, but with uffd the entire memory region is read/write) when faulted and then explicitly changes the protection level of the page to `NONE` to induce a `SIGSEGV` when the kernel retries the fault, ultimately resulting in a trap as if the page were `NONE` the whole time.

As an instance may continue to be used after a trap or (with the pooling allocator) the same virtual memory pages reused for a new instance, we have to reset the protection level of these "none" pages back to read-write because they might actually be accessible pages in the future if a memory grows (this should probably check to see if the page is beyond the limit of addressable pages and skip recording it if so, though; I'll fix that).

The ultimate goal of the `uffd` feature is to (almost) never change memory protection levels as doing so requires taking kernel locks.